src/Connector.sol

```diff
@@ -126,75 +160,115 @@ contract CentrifugeConnector {
-        // TODO(nuno)
+    function decreaseRedeemOrder(uint64 poolId, bytes16 trancheId, address currencyAddress, uint128 amount) public {
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        uint128 currency = currencyAddressToId[currencyAddress];
+        require(currency != 0, "CentrifugeConnector/unknown-currency");
+        require(poolCurrencies[poolId][currencyAddress], "CentrifugeConnector/pool-currency-not-allowed");
+
+        gateway.decreaseRedeemOrder(poolId, trancheId, msg.sender, currency, amount);
     }
 
     function collectRedeem(uint64 poolId, bytes16 trancheId) public {
-        // TODO(nuno)
-    }
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
 
-    function collectForRedeem(uint64 poolId, bytes16 trancheId, bytes32 userAddress) public {
-        // TODO(nuno)
+        gateway.collectRedeem(poolId, trancheId, address(msg.sender));
     }
 
     function collectInvest(uint64 poolId, bytes16 trancheId) public {
-        // TODO(nuno)
-    }
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
 
-    function collectForInvest(uint64 poolId, bytes16 trancheId, bytes32 userAddress) public {
-        // TODO(nuno)
+        gateway.collectInvest(poolId, trancheId, address(msg.sender));
     }
 
     // --- Incoming message handling ---
-    // todo(nuno): store currency and decimals
-    function addPool(uint64 poolId, uint128 currency, uint8 decimals) public onlyGateway {
+    function addCurrency(uint128 currency, address currencyAddress) public onlyGateway {
+        require(currencyIdToAddress[currency] == address(0), "CentrifugeConnector/currency-already-added");
+        require(currencyAddressToId[currencyAddress] == 0, "CentrifugeConnector/currency-already-added");
+
+        currencyIdToAddress[currency] = currencyAddress;
+        currencyAddressToId[currencyAddress] = currency;
+        emit CurrencyAdded(currency, currencyAddress);
+    }
+
+    function addPool(uint64 poolId) public onlyGateway {
         Pool storage pool = pools[poolId];
         require(pool.createdAt == 0, "CentrifugeConnector/pool-already-added");
         pool.poolId = poolId;
         pool.createdAt = block.timestamp;
         emit PoolAdded(poolId);
     }
 
+    function allowPoolCurrency(uint128 currency, uint64 poolId) public onlyGateway {
```

Comment:
This is definitely nit picking and I realize it would require changes on both sides, but, semantically wouldn't it make more sense if in the message and transaction parameters, `poolId` was before `currency`?

Similar to how if you are investing X in pool Y, then the arguments would be (pool Y, amount X).
And here you are allowing currency X for pool Y, so the arguments make sense to be (pool Y, currency X).

Reply:
I also concluded this because the naming of the event was `PoolCurrencyAllowed`, as `CurrencyPoolAllowed` does not make sense.

Reply:
Yeah that makes sense, we should do that

Reply:
Actually, the reason why I chose this order was that it reads "AllowPoolCurrency => allow currency C for pool P" instead of "allow, for pool P, currency C". But indeed for the investing messages we have poolId before so I am cool with keeping that format

Reply:
Done and updated the spec